asm/instruction.go

```diff
@@ -35,17 +35,36 @@ type Instruction struct {
 	Offset   int16
 	Constant int64
 
-	// Reference denotes a reference (e.g. a jump) to another symbol.
-	Reference string
+	// Metadata contains optional metadata about this instruction
+	metadata *metadata
+}
+
+// WithSymbol adds symbol metadata to the instruction, which other instructions can use as Reference.
```

Comment:
The metadata part is an implementation detail and can be omitted. We should attempt to accurately describe the DSL instead.

```suggestion
// WithSymbol marks the Instruction as a Symbol, which other Instructions
// can point to using corresponding calls to WithReference.
```